use_cases/fetcher.go

```diff
@@ -19,26 +21,65 @@ type Fetcher struct {
 	storage writer
 }
 
+type result struct {
+	Error   error
+	Pokemon models.Pokemon
+}
+
 func NewFetcher(api api, storage writer) Fetcher {
 	return Fetcher{api, storage}
 }
 
 func (f Fetcher) Fetch(from, to int) error {
 	var pokemons []models.Pokemon
-	for id := from; id <= to; id++ {
-		pokemon, err := f.api.FetchPokemon(id)
-		if err != nil {
-			return err
-		}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
 
-		var flatAbilities []string
-		for _, t := range pokemon.Abilities {
-			flatAbilities = append(flatAbilities, t.Ability.URL)
+	ch := f.generator(ctx, from, to)
+	for result := range ch {
+		if result.Error != nil {
+			return result.Error
 		}
-		pokemon.FlatAbilityURLs = strings.Join(flatAbilities, "|")
-
-		pokemons = append(pokemons, pokemon)
+		pokemons = append(pokemons, result.Pokemon)
 	}
 
 	return f.storage.Write(pokemons)
 }
+
+func (f Fetcher) generator(ctx context.Context, from, to int) <-chan result {
```

Comment:
even though this is called generator, this is more sort of fan in/ fan out pattern as you are separating the workloads into workers and then waiting for a response in the channel, why are you calling this generator?

Reply:
I used the code from the workshop as a reference. https://github.com/davidcdorbecker/workshop-4/blob/master/generator/generator.go